### PR TITLE
added nsurlsession instrumentation

### PR DESCRIPTION
**Pull request by bryce-b**
pretty messy first pass; it will need some clean up, but I wanted to get this pr up for comment.

### CODE REVIEW

File: Sources/Instrumentation/NSURLSession/DDURLSessionDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+@objc
+open class DDURLSessionDelegate: NSObject, URLSessionTaskDelegate {
+    var interceptor: URLSessionInterceptorType?
+
+    @objc
+    override public init() {
+        interceptor =   URLSessionAutoInstrumentation.instance?.interceptor
+        if interceptor == nil {
+            print("""
+                `Agent.start()` must be called before initializing the `DDURLSessionDelegate` and
```
Comment (bryce-b) [inline Sources/Instrumentation/NSURLSession/DDURLSessionDelegate.swift:22]: review messaging here

**Comment by nachoBonafonte** [comment]
My main concern with this approach is the need for the user to use the provided session delegate, so users will need to modify their code in order to use this instrumentation, they may already have a session delegate that should inherit from the provided one. If application uses some library that does networking it could be impossible to instrument this third party library without modifying the code.

**Comment by bryce-b** [comment]
It should be pretty straight forward to add some instrumentation to intercept the delegate when it is passed to the NSURLSession. It's probably a good idea to make a couple of options available for users, which ever they are.

**Comment by nachoBonafonte** [comment]
I have added my proposal also as a PR: [here](https://github.com/open-telemetry/opentelemetry-swift/pull/119)

**Pull request closed without merging**